js/style/style.js

```diff
@@ -318,13 +348,13 @@ Style.prototype = util.inherit(Evented, {
         if (this.sources[id] !== undefined) {
             throw new Error('There is already a source with this ID');
         }
-        if (!Source.is(source) && this._handleErrors(validateStyle.source, 'sources.' + id, source)) return this;
+        var builtIns = ['vector', 'raster', 'geojson', 'video', 'image'];
+        var shouldValidate = !Source.is(source) && builtIns.indexOf(source.type) >= 0;
+        if (shouldValidate && this._handleErrors(validateStyle.source, 'sources.' + id, source)) return this;
 
-        source = Source.create(source);
+        source = new TilePyramid(id, source, this.dispatcher);
```

Comment:
This would be fixed if we decide to go ahead with https://github.com/mapbox/mapbox-gl-js/pull/2667/files/3d6cf0a5ce5d118c52d2fe4446daffad25dd79b2#r65790830